---
title: "Practicals 01: Sequence search (intro)"
subtitle: "BE_22 Bioinformatics SS 22"
author: "January Weiner"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float:
        collapsed: true
        smooth_scroll: true
    toc_depth: 3
    fig_caption: yes
    number_sections: true
abstract: |
---

```{r,echo=FALSE}
## Set default options for the knitr RMD processing
knitr::opts_chunk$set(echo=FALSE,warning=FALSE,message=FALSE,fig.width=5,fig.height=5,cache=TRUE,autodep=TRUE, results="hide")
```

```{r libraries,cache=FALSE}
```

# Self-check

 * What is a sequence?
 * What is a 5' end? a 3' end? N- terminus? C- terminus?
 * What is a complementary sequence?
 * What are the differences between RNA and DNA sequences?

# Alignments and dot plots

What is an alignment? How does it differ from a complementary sequence?
What is a local alignment? What is a global alignment?

## Resources

For the dot plots and alignments you will be using EMBOSS. EMBOSS is a
suite of command line tools for bioinformatics, you can download it from
[here](http://emboss.sourceforge.net/download/). If you are brave, feel
free to download and use it from command line. Alternatively, download
[Jemboss](http://emboss.sourceforge.net/Jemboss/) and use the GUI Java
interface.

However, many sites offer
a web interface, which is the way I will show you these tools in the
course.

Links to online interfaces (use the one that you like)

 * bioinformatics.nl: [https://www.bioinformatics.nl/cgi-bin/emboss/](https://www.bioinformatics.nl/cgi-bin/emboss/)
 * (slower) European Bioinformatic Institute: [https://www.ebi.ac.uk/Tools/emboss/](https://www.ebi.ac.uk/Tools/emboss/)

[EMBOSS reference](https://www.cell.com/trends/genetics/fulltext/S0168-9525(00)02024-2): *Rice, Peter, Ian Longden, and Alan Bleasby. "EMBOSS: the
European molecular biology open software suite." Trends in genetics 16.6
(2000): 276-277.*

**Note:** in previous courses, on rare occasion it happened that the sites
noticed mutiple connections from one IP range and the firewall
automatically shut us down. This should not happen; if it does, we will
have to waste time on installing Jemboss. However, installing unfamiliar
tools is a large portion of life in bioinformatics, so that will also be a
valuable exercise.

## Dot plots

  Use the "dotmatcher" program from EMBOSS.

  The program does not merely put a dot on the plot for each match.
  Instead, the matches are averaged over a number of nucleotides or amino
  acid residues, and the result is then compared with a threshold value
  (only values above the threshold are then displayed).

 1. Create a dot plot for the following two sequences. This is the FASTA
    format â€“ sequences are stored in a simple text file; each sequence is
    preceded with a line starting with a "greater than" sign `>`. These
    sequences are very similar, you should have no issues seeing the dot
    plot.

```
>sequence1
ACTCTTCTGGTCCCCACAGACTCAGAGAGAACCCACCATGGTGCTGTCTCCTGCCGACAAGACCAACGTC
AAGGCCGCCTGGGGTAAGGTCGGCGCGCACGCTGGCGAGTATGGTGCGGAGGCCCTGGAGAGGTGAGGCT
CCCTCCCCTGCTCCGACCCGGGCTCCTCGCCCGCCCGGACCCACAGGCCACCCTCAACCGTCCTGGCCCC
GGACCCAAACCCCACCCCTCACTCTGCTTCTCCCCGCAGGATGTTCCTGTCCTTCCCCACCACCAAGACC
TACTTCCCGCACTTCGACCTGAGCCACGGCTCTGCCCAGGTTAAGGGCCACGGCAAGAAGGTGGCCGACG
CGCTGACCAACGCCGTGGCGCACGTGGACGACATGCCCAACGCGCTGTCCGCCCTGAGCGACCTGCACGC
GCACAAGCTTCGGGTGGACCCGGTCAACTTCAAGGTGAGCGGCGGGCCGGGAGCGATCTGGGTCGAGGGG
CGAGATGGCGCCTTCCTCGCAGGGCAGAGGATCACGCGGGTTGCGGGAGGTGTAGCGCAGGCGGCGGCTG
CGGGCCTGGGCCCTCGGCCCCACTGACCCTCTTCTCTGCACAGCTCCTAAGCCACTGCCTGCTGGTGACC
CTGGCCGCCCACCTCCCCGCCGAGTTCACCCCTGCGGTGCACGCCTCCCTGGACAAGTTCCTGGCTTCTG
TGAGCACCGTGCTGACCTCCAAATACCGTTAAGCTGGAGCCTCGGTGGCCATGCTTCTTGCCCCTTGGGC
CTCCCCCCAGCCCCTCCTCCCCTTCCTGCACCCGTACCCCCGTGGTCTTTGAATAAAGTCTGAGTGGGCG
GCA
```

```
>sequence2
AGACTCTGGCCAACACCCCCTGTAAGGCCACAGGAGAGGAACAGGAGTGATAGCCCCCAAACCCCAGTCC
CACCAGGCCCTGAGGGCCCCTTTGTCACTGGATCTGATAAGAAACACCACCCCTGCAGCCCCCTCCCCTC
ACCTGACCAATGGCCACAGCCTGGCTGGGCCCAGCTCCCTGTATATAAGGGGACCCTGGGGGCTGAGCAC
TACCAAGGCCAGTCCTGAGCAGGCCCAACTCCAGTGCAGCTGCCCACCCTGCCGCCATGTCTCTGACCAA
GACTGAGAGGACCATCATTGTGTCCATGTGGGCCAAGATCTCCACGCAGGCCGACACCATCGGCACCGAG
ACTCTGGAGAGGCTCTTCCTCAGCCACCCGCAGACCAAGACCTACTTCCCGCACTTCGACCTGCACCCGG
GGTCCGCGCAGTTGCGCGCGCACGGCTCCAAGGTGGTGGCCGCCGTGGGCGACGCGGTGAAGAGCATCGA
CGACATCGGCGGCGCCCTGTCCAAGCTGAGCGAGCTGCACGCCTACATCCTGCGCGTGGACCCGGTCAAC
TTCAAGCTCCTGTCCCACTGCCTGCTGGTCACCCTGGCCGCGCGCTTCCCCGCCGACTTCACGGCCGAGG
CCCACGCCGCCTGGGACAAGTTCCTATCGGTCGTATCCTCTGTCCTGACCGAGAAGTACCGCTGAGCGCC
GCCTCCGGGACCCCCAGGACAGGCTGCGGCCCCTCCCCCGTCCTGGAGGTTCCCCAGCCCCACTTACCGC
GTAATGCGCCAATAAACCAATGAACGAA
```

 2. Play with the parameters "window size" and "threshold". Increase and
    decrease the threshold: what happens? What happens if you set the
    threshold to 1 and windows size to 3? 

 3. Use the window "20". Try increasing the threshold until you almost see
    no dots. What is the threshold value? What is the difference between
    the two sequences? What are the similarities?

 3. Modified the sequences: cut out a fragment from the middle of
    one of the sequences and display the dot plot. Do you see what happens?
    Now copy and paste a fragment two or more times. Do you see what happens?


## Global and local alignments

You will create global alignments with the NW algorithm (program "needle")
and the local alignments with the Smith-Waterman algorithm (program
"water").

 1. Create global and local alignments between the above sequences 1 and 2 as well as
    3 and 4.
 
 2. The following are the protein sequences corresponding to sequences 1
    and 2. Run the alignments. Any thoughts?

```
>sequence1_prot
MVLSPADKTNVKAAWGKVGAHAGEYGAEALERMFLSFPTTKTYFPHFDLSHGSAQVKGHG
KKVADALTNAVAHVDDMPNALSALSDLHAHKLRVDPVNFKLLSHCLLVTLAAHLPAEFTP
AVHASLDKFLASVSTVLTSKYR
```

```
>sequence2_prot
MSLTKTERTIIVSMWAKISTQADTIGTETLERLFLSHPQTKTYFPHFDLHPGSAQLRAHG
SKVVAAVGDAVKSIDDIGGALSKLSELHAYILRVDPVNFKLLSHCLLVTLAARFPADFTA
EAHAAWDKFLSVVSSVLTEKYR
```

 3. Cut out a fragment from the middle of one of the sequences. Run the
    alignment and take a look. Now set the gap opening penalty and gap
    extension penalty to 1. What is the difference? How about 25/0.5?

 4. (hard) There is a consipracy theory online that the spike protein from
    Sars-Cov-2 is similar to the human syncytin-1 protein (which is of viral
    origin). Find the two proteins in the Uniprot database (`SPIKE_SARS`
    and and create global and local alignments and dot plots. Can you see
    any similarity? Try different scoring matrices (BLOSUM30, BLOSUM62
    etc.) and different gap penalties.

# BLAST searches

Like often in  bioinformatics, the blast family of programs can be used as
command line programs. You can download and install them on your computer
(in Ubuntu, it is as simple as `sudo apt install ncbi-blast+`); similarly,
you can download and install the NCBI databases such as NR or refseq
locally. This is advantageous when you have more than just a few sequences
to blast.

However, BLAST features a popular and extremly refined user interface on
the NCBI website, which we will use today. The entry point is
[https://blast.ncbi.nlm.nih.gov/Blast.cgi](https://blast.ncbi.nlm.nih.gov/Blast.cgi).

 1. Using blastn, identify the sequences 1, 2, 3 and 4.


 2. Michael Crichton included a DNA sequence in his book "Lost World". Use
BLAST to identify the sequence origin. You might try to limit the search to
a group of organisms.

```
>DinoDNA  "Dinosaur DNA" from Crichton LOST WORLD
GAATTCCGGAAGCGAGCAAGAGATAAGTCCTGGCATCAGATACAGTTGGAGATAAGGACG
GACGTGTGGCAGCTCCCGCAGAGGATTCACTGGAAGTGCATTACCTATCCCATGGGAGCC
ATGGAGTTCGTGGCGCTGGGGGGGCCGGATGCGGGCTCCCCCACTCCGTTCCCTGATGAA
GCCGGAGCCTTCCTGGGGCTGGGGGGGGGCGAGAGGACGGAGGCGGGGGGGCTGCTGGCC
TCCTACCCCCCCTCAGGCCGCGTGTCCCTGGTGCCGTGGGCAGACACGGGTACTTTGGGG
ACCCCCCAGTGGGTGCCGCCCGCCACCCAAATGGAGCCCCCCCACTACCTGGAGCTGCTG
CAACCCCCCCGGGGCAGCCCCCCCCATCCCTCCTCCGGGCCCCTACTGCCACTCAGCAGC
GGGCCCCCACCCTGCGAGGCCCGTGAGTGCGTCATGGCCAGGAAGAACTGCGGAGCGACG
GCAACGCCGCTGTGGCGCCGGGACGGCACCGGGCATTACCTGTGCAACTGGGCCTCAGCC
TGCGGGCTCTACCACCGCCTCAACGGCCAGAACCGCCCGCTCATCCGCCCCAAAAAGCGC
CTGCGGGTGAGTAAGCGCGCAGGCACAGTGTGCAGCCACGAGCGTGAAAACTGCCAGACA
TCCACCACCACTCTGTGGCGTCGCAGCCCCATGGGGGACCCCGTCTGCAACAACATTCAC
GCCTGCGGCCTCTACTACAAACTGCACCAAGTGAACCGCCCCCTCACGATGCGCAAAGAC
GGAATCCAAACCCGAAACCGCAAAGTTTCCTCCAAGGGTAAAAAGCGGCGCCCCCCGGGG
GGGGGAAACCCCTCCGCCACCGCGGGAGGGGGCGCTCCTATGGGGGGAGGGGGGGACCCC
TCTATGCCCCCCCCGCCGCCCCCCCCGGCCGCCGCCCCCCCTCAAAGCGACGCTCTGTAC
GCTCTCGGCCCCGTGGTCCTTTCGGGCCATTTTCTGCCCTTTGGAAACTCCGGAGGGTTT
TTTGGGGGGGGGGCGGGGGGTTACACGGCCCCCCCGGGGCTGAGCCCGCAGATTTAAATA
ATAACTCTGACGTGGGCAAGTGGGCCTTGCTGAGAAGACAGTGTAACATAATAATTTGCA
CCTCGGCAATTGCAGAGGGTCGATCTCCACTTTGGACACAACAGGGCTACTCGGTAGGAC
CAGATAAGCACTTTGCTCCCTGGACTGAAAAAGAAAGGATTTATCTGTTTGCTTCTTGCT
GACAAATCCCTGTGAAAGGTAAAAGTCGGACACAGCAATCGATTATTTCTCGCCTGTGTG
AAATTACTGTGAATATTGTAAATATATATATATATATATATATATCTGTATAGAACAGCC
TCGGAGGCGGCATGGACCCAGCGTAGATCATGCTGGATTTGTACTGCCGGAATTC
```

 2. Now use `blastx`, which takes a nucleotide sequence as input but
    searches in protein sequence databases. What do you see? Can you see
    the hidden message in the protein sequence? Identify the nucleotides
    corresponding to the hidden message.

 3. Using appropriate alignment tool, create a global alignment between the
    Crichton sequence and the best hit. How can you get FASTA from the
    blast output? How can you get a FASTA nucleotide sequence?

 4. Here is another one, this time from "Jurassic Park". Where does the
    sequence come from?

```
>DinoDNA  "Dinosaur DNA" from Crichton JURASSIC PARK  p. 103 nt 1-1200
GCGTTGCTGGCGTTTTTCCATAGGCTCCGCCCCCCTGACGAGCATCACAAAAATCGACGC
GGTGGCGAAACCCGACAGGACTATAAAGATACCAGGCGTTTCCCCCTGGAAGCTCCCTCG
TGTTCCGACCCTGCCGCTTACCGGATACCTGTCCGCCTTTCTCCCTTCGGGAAGCGTGGC
TGCTCACGCTGTACCTATCTCAGTTCGGTGTAGGTCGTTCGCTCCAAGCTGGGCTGTGTG
CCGTTCAGCCCGACCGCTGCGCCTTATCCGGTAACTATCGTCTTGAGTCCAACCCGGTAA
AGTAGGACAGGTGCCGGCAGCGCTCTGGGTCATTTTCGGCGAGGACCGCTTTCGCTGGAG
ATCGGCCTGTCGCTTGCGGTATTCGGAATCTTGCACGCCCTCGCTCAAGCCTTCGTCACT
CCAAACGTTTCGGCGAGAAGCAGGCCATTATCGCCGGCATGGCGGCCGACGCGCTGGGCT
GGCGTTCGCGACGCGAGGCTGGATGGCCTTCCCCATTATGATTCTTCTCGCTTCCGGCGG
CCCGCGTTGCAGGCCATGCTGTCCAGGCAGGTAGATGACGACCATCAGGGACAGCTTCAA
CGGCTCTTACCAGCCTAACTTCGATCACTGGACCGCTGATCGTCACGGCGATTTATGCCG
CACATGGACGCGTTGCTGGCGTTTTTCCATAGGCTCCGCCCCCCTGACGAGCATCACAAA
CAAGTCAGAGGTGGCGAAACCCGACAGGACTATAAAGATACCAGGCGTTTCCCCCTGGAA
GCGCTCTCCTGTTCCGACCCTGCCGCTTACCGGATACCTGTCCGCCTTTCTCCCTTCGGG
CTTTCTCAATGCTCACGCTGTAGGTATCTCAGTTCGGTGTAGGTCGTTCGCTCCAAGCTG
ACGAACCCCCCGTTCAGCCCGACCGCTGCGCCTTATCCGGTAACTATCGTCTTGAGTCCA
ACACGACTTAACGGGTTGGCATGGATTGTAGGCGCCGCCCTATACCTTGTCTGCCTCCCC
GCGGTGCATGGAGCCGGGCCACCTCGACCTGAATGGAAGCCGGCGGCACCTCGCTAACGG
CCAAGAATTGGAGCCAATCAATTCTTGCGGAGAACTGTGAATGCGCAAACCAACCCTTGG
CCATCGCGTCCGCCATCTCCAGCAGCCGCACGCGGCGCATCTCGGGCAGCGTTGGGTCCT
```

 5. (extra) Use the Sars-Cov-2 spike protein sequence (see above) to search
    for similar proteins in the human proteome. Which blast program would
    you choose? What BLOSUM matrix would you choose given that one expects
    only very distant relationships? How do you limit the search to human
    proteins? How do you change the parameters to see even very remote
    relationships? Is syncytin one of the hits?

## Homework

Here is a mystery sequence. What is this protein? From which organism does it
come from? (X stands for an unknown amino acid).

```
>mystery sequence
GATGAPGIAGAPGFPGARGAPGPQGPSGAPGPKXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXGVQGPPGPQGPR
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXGSAGPPGATGFP
GAAGRXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXGVVGLPGQR
```

Identify the sequence. Use the blastp program and the NR database. There
are now several similar sequences in the NR database, use the one that
comes from an organism called "Tyr....". Click on the "Sequence ID" link to
view more information about the sequence. and find and read the paper from
which the mystery sequence is derived. Provide a five sentence (roughly)
summary of what has been done and how and what the conclusions are.


