<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Practicals 01: Sequence search (intro)</title>
    <meta charset="utf-8" />
    <meta name="author" content="January Weiner" />
    <meta name="date" content="2022-03-24" />
    <script src="libs/header-attrs-2.11/header-attrs.js"></script>
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link rel="stylesheet" href="files/cubi-fonts.css" type="text/css" />
    <link rel="stylesheet" href="files/style.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Practicals 01: Sequence search (intro)
## BE_22 Bioinformatics SS 21
### January Weiner
### 2022-03-24

---






---


## Sequence alignments ✪

Examples of alingments of sequences `ACTTGA` and `ACCTGGA`

```
ACTTG-A    ACTT-GA     AC-TTG-A    ---ACTTGA    ACTTGA-------
|| || |    || | ||     || | | |         | ||
ACCTGGA    ACCTGGA     ACCT-GGA    ACC--TGGA    ------ACCTGGA
```

This is not a correct alignment (but may make sense in a different context):

```
ACTTG--A
|| ||  |
ACCTG-GA
```

---

## Do not confuse alignments with complementary sequences

.pull-left[
Alignment:

```
5'-ACTTG-A-3'
   || || |
5'-ACCTGGA-3'
```
]

.pull-right[
Complementary sequence:

```
5'-ACTTGGA-3'
   |||||||
3'-TGAACCT-5'
```
]

---

## Dot plots ✪

```
                         

        A C G G T A      
      A *         *      
      C   *              
      G     * *          
      T         *        
      A *         *      
                         
                         
                         
                         
                                                         
```


---

## Dot plots

```
                         Duplication        

        A C G G T A        A C G G T A      
      A *         *      A *         *      
      C   *              C   *              
      G     * *          G     * *          
      T         *        T         *        
      A *         *      A *         *      
                         C   *                            
                         G     * *                        
                         T         *                      
                         A *         *                    
                                                         
```


---

## Dot plots

```
                         Duplication        Insertion / Deletion

        A C G G T A        A C G G T A        A C G G G G G G T A 
      A *         *      A *         *      A *                 * 
      C   *              C   *              C   *                 
      G     * *          G     * *          G     * * * * * *     
      T         *        T         *        T                 *   
      A *         *      A *         *      A *                 * 
                         C   *                            
                         G     * *                        
                         T         *                      
                         A *         *                    
                                                         
```


---

## The BLAST algorithm

BLAST – Basic Local Alignment Search Tool

![](images/blast.png)

![](images/blast_2.png)

---

## The BLAST algorithm

![](images/blast_algorithm.jpg)

---

## The BLAST algorithm


 * Advantage: `\(O(n)\)` – linear time complexity
 * Compromise between speed and sensitivity
 * Heuristic, not exact
 * Primary output: HSP, high scoring segment pairs (possibly multiple per
   sequence pair)

---

## Important BLAST parameters

 * Type of BLAST (which program)
 * Database (more on that tomorrow)
 * Word size (lower word size = slower but more sensitive)
 * Filter low complexity regions (e.g. repeats)

---

## BLAST E-value

 * Expected number of HSP's which have a score equal or better to the given
   result
 
  `\(E = m\cdot n\cdot 2^{-S'}\)`

  where `\(S'\)` is a normalized score and `\(m\)` and `\(n\)` are sequence lengths.

---

## BLAST programs


|Program |Description                                   |Input      |Database   |
|:-------|:---------------------------------------------|:----------|:----------|
|blastn  |Search sequence against sequence              |Nucleotide |Nucleotide |
|blastp  |Search sequence against sequence              |Protein    |Protein    |
|blastx  |Translate query in all open reading frames    |Nucleotide |Protein    |
|tblastn |Translate database in all open reading frames |Protein    |Nucleotide |

---

## Databases

 * [NCBI Genbank](https://www.ncbi.nlm.nih.gov/genbank/)
    * Refseq (reference sequences, curated)
    * nr – Non-redundant (almost all known sequences)
    * Environmental sequences
    * Entrez Gene – gene centered, provides Entrez identifiers
 * [SwissProt / Uniprot](https://uniprot.org) – protein sequences
 * [PDB - Protein Databank](https://www.wwpdb.org) – protein structure database
 * [Ensembl](https://ensembl.org) – Ensembl genome browser and database
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script src="files/macros.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"countIncrementalSlides": false,
"ratio": "16:9"
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
