---
title: "Lecture 11: Gene set enrichments"
subtitle: "BE_22 Bioinformatics SS 21"
author: "January Weiner"
date: "`r Sys.Date()`"
outputdupa: html_document
output:
  xaringan::moon_reader:
    self-contained: true
    css: ["default", "files/cubi-fonts.css", "files/style.css" ]
    lib_dir: libs
    nature:
      highlightStyle: github
      countIncrementalSlides: false
      ratio: "16:9"
      beforeInit: "files/macros.js"
toc: no
---

```{r,echo=FALSE}
## Set default options for the knitr RMD processing
knitr::opts_chunk$set(echo=FALSE,warning=FALSE,message=FALSE,fig.width=5,fig.height=5,cache=FALSE,autodep=TRUE, results="hide")
library(pander)
library(knitr)
library(kableExtra)
#source("functions.R")
```

```{r echo=FALSE,include=FALSE,eval=TRUE}
options(crayon.enabled = TRUE)
options(crayon.colors = 256)
knitr::knit_hooks$set(output = function(x, options){
  paste0(
    '<pre class="r-output"><code>',
    fansi::sgr_to_html(x = htmltools::htmlEscape(x), warn = FALSE),
    '</code></pre>'
  )
})

## this is an ugly, ugly hack, but otherwise crayon does not LISTEN TO REASON!!!
num_colors <- function(forget=TRUE) 256
library(crayon)
assignInNamespace("num_colors", num_colors, pos="package:crayon")
```


```{r libraries,cache=FALSE}
library(tidyverse)
library(ggplot2)
library(readxl)
library(cowplot)
library(RColorBrewer)
library(cluster)
library(pca3d)
library(Rseasnap)
library(colorDF)
library(DESeq2)
library(tmod)
```

<!--
class:empty-slide,myinverse
background-image:url(images/arnolfini.jpg)
-->



## Gene set enrichment analysis

```{r}
covar <- readRDS("../../Datasets/covar.rds")
cnts  <- readRDS("../../Datasets/counts.rds")
annot <- readRDS("../../Datasets_not_mirrored/annot.rds")
```

```{r deseq2,cache=TRUE}
library(DESeq2)
ds2 <- DESeqDataSetFromMatrix(countData = cnts, colData=covar, design = ~group)
ds2 <- DESeq(ds2)
res <- map(rlang::set_names(resultsNames(ds2)), ~ results(ds2, name=.x))
```

The problem: how do you interpret a list of 1000 results?

---

```{r results="asis"}
res <- map(res, ~ .x %>% as.data.frame %>% arrange(padj) %>% { merge(annot, ., by.x="PrimaryID", by.y=0)})
DT::datatable(res[[2]] %>% select(SYMBOL, GENENAME, log2FoldChange, padj)) %>%
  DT::formatSignif(c("log2FoldChange", "padj"), digits=2)
```


---
class:empty-slide,myinverse
background-image:url(images/desert2.jpg)

.mytop[
The number of grains of sand in this picture accurately represents the
number of approaches for functional analysis of gene expression.
]

---

## Plan for today

 * Gene set enrichment analysis
    * 1st generation: hypergeometric test, Fisher's exact test
    * 2nd generation: other algorithms
    * 3rd generation: pathway topology analysis


---

## Gene ontology database

.pull-left[

 * Genes are assigned to GO categories
 * GO categories are assigned to each other
 * GO can be represented as a *directed acyclic graph*

]

.pull-right[
![](images/gorilla.png)
]

---

## KEGG pathways

KEGG – Kyoto Encyclopaedia of Genes and Genomes

---
class:empty-slide,myinverse
background-image:url(images/hsa04110.gse16873.png)

---


## Other gene sets

Different lists of genes can be obtained from various sources.

 * Coexpression data sets: groups of genes that have correlated expression
   profiles

--

 * Results from another differential gene expression analysis experiment:
   groups of genes significant in a comparison

--

 * Targets of a particular miRNA determined computationally


---

## MSigDB – a meta database of gene sets

.pull-left[
![:scale 60%](images/msigdb.png)
]

---

## Group of genes, and then what?

First, define DEGs: differentially expressed genes (e.g.
`abs(log2FoldChange) > 1`, `adj_p_val < 0.001`)

Second, define which genes are in a given gene set (GS), and which are not.

.pull-left[

**No significant enrichment:**

|    |DEGs |non-DEGs|Fraction|
|----|-----|--------|--------|
|in GS|5 | 95| 5%|
|non in GS|50 |1000|5%|
|Fraction|9%|9.5%||

]

--

.pull-right[


**Significant enrichment:**

|    |DEGs |non-DEGs|Fraction|
|----|-----|--------|--------|
|in GS|50 | 60| 45%|
|non in GS|70 |1000|7%|
|Fraction|42%|6%||

]

How can we test it?

---

.pull-left[

**No significant enrichment:**

|    |DEGs |non-DEGs|Fraction|
|----|-----|--------|--------|
|in GS|5 | 95| 5%|
|non in GS|50 |1000|5%|
|Fraction|9%|9.5%||


]

--

.pull-right[

**Significant enrichment:**

|    |DEGs |non-DEGs|Fraction|
|----|-----|--------|--------|
|in GS|50 | 60| 45%|
|non in GS|70 |1000|7%|
|Fraction|42%|6%||

]

--

.pull-left[

Results of $\chi^2$ test:

Left:

```{r, results="markdown"}
chisq.test(matrix(c(5, 95, 50, 1000), byrow=TRUE, nrow=2))
```

Right:


```{r, results="markdown"}
chisq.test(matrix(c(50, 60, 70, 1000), byrow=TRUE, nrow=2))
```

]


---

## Disadvantages of the generation 1 approach

 * You have to define DEGs... but that depends on the number of samples
    * the larger the number of samples, the higher statistical power
    * the higher statistical power, the more DEGs you can detect
    * therefore, the results depend on an arbitrary threshold

 * What to do if you have no DEGs at all?

---

## Generation 2 approaches – "Functional Class Sorting"

 * do not depend on arbitrary thresholds

---

## GSEA algorithm

![](images/GSEA-homegraphic.gif)


---

## tmod algorithm

 * Arrange the genes according to p-value or another similar metrics
 * Test whether genes from a given gene set are denser on the beginning of
   the list

---
class:empty-slide,myinverse
background-image:url(images/img21.png)

---
class:empty-slide,myinverse
background-image:url(images/img22.png)

---
class:empty-slide,myinverse
background-image:url(images/img23.png)

---
class:empty-slide,myinverse
background-image:url(images/img24.png)

---
class:empty-slide,myinverse
background-image:url(images/img25.png)

---
class:empty-slide,myinverse
background-image:url(images/img26.png)

---
class:empty-slide,myinverse
background-image:url(images/img27.png)

---
class:empty-slide,myinverse
background-image:url(images/img28.png)

---

### Tmod statistics

Since normalized ranking for a gene $i$

$r'_{i}=\frac{r_i}{N} \sim U(0,1)$

--

then

$-\ln{r'_{i}} \sim Exp(1)$

--

and

$-2\cdot\ln{r'_i} \sim \chi^2_2$

--

thus, if $G$ is a set of genes,

$-2\cdot\sum_{i \in G}\ln{r'_i} \sim \sum_{i \in G}\chi^2_2 \sim \chi^2_{2\cdot n(G)}$

This gives a fast and reliable statistics with a known distribution which
can be used to associate a p-value to any set of genes.


---

![:scale 90%](images/tmod.png)

.footnote[*Zyla J, Marczyk M, Domaszewska T, Kaufmann SH, Polanska J, Weiner
3rd J. Gene set enrichment for reproducible science: comparison of CERNO
and eight other algorithms. Bioinformatics. 2019 Dec 15;35(24):5146-54.*]

---

![:scale 90%](images/tmod_2.png)

---

![:scale 90%](images/tmod_3.png)

.pull-left[Vaccine with an adjuvant]

.pull-right[Vaccine without an adjuvant]

.footnote[*Weiner, January, et al. "Characterization of potential biomarkers
of reactogenicity of licensed antiviral vaccines: randomized controlled
clinical trials conducted by the BIOVACSAFE consortium." Scientific reports
9.1 (2019): 1-14.*]


---

## Network analysis (generation 3)

Genes / proteins may be linked by

 * co-expression 
 * common pathways
 * literature (co-citation)
 * synteny (co-occurence in genomes)
 * common motifs

---

## STRING

![](images/string.png)


